### ISSUE TITLE
Querying via advanced query or endpoints

### ISSUE THREAD

**Issue by jasloe** [jasloe/archivesspace-drupal#23]
@seth-shaw-unlv: I have sort of a big picture question. 

Which approach is more appropriate? Advanced queries against the search feed (and also the delete feed, if we're feeling ambitious)? Something like....

<pre>
search?page=1&?page_size=1&aq={"query":{"op":"AND","subqueries":[{"field":"instance_type_enum_s","archvial_object":"item","comparator":"equals","jsonmodel_type":"field_query"},{"field":"user_mtime","value":"2018-08-07","comparator":"greater_than","jsonmodel_type":"date_field_query"}],"jsonmodel_type":"boolean_query"},"jsonmodel_type":"advanced_query"}'
</pre>

In this case, we'd collect search criteria in a form, package that up, ship it off to ASpace, and parse the resulting JSON blob. 

That's the way we did it in D7, though the advanced query interface is largely undocumented except for the list of field properties in the Solr schema config file (I think). This would give the user front-end tools to refine queries, by date for example. 

Or target specific endpoints the usual way, e.g. return archival_objects?id_set=123,456,789 and parse the results. This is much simpler, obviously. 

Maybe I should take a look at some of the other Python clients and see what other folks have done....

Thoughts?

**Comment by seth-shaw-unlv** [comment]
I was thinking we would use the API to slurp in everything that is published. Start with Repositories (/repositories to get a list and then /repositories/:id for each) then go on to Agents, Subjects, Archival Objects, and Resources. (Although my testing would start with Repositories and Resources.)

**Comment by seth-shaw-unlv** [comment]
Of course, that approach isn't scalable for recurring updates, because you have to hit each object to see if it has changed. Perhaps the advanced query with mtime would be better. We would have to track a "last queried" field somewhere.

**Comment by jasloe** [comment]
Maybe write an agent that notifies Drupal of updates? Changes to lock version for example?

**Comment by seth-shaw-unlv** [comment]
Well, if we are relying on the Migrate API it is simply a "go do an update". The source plugin would need to be able to query ArchivesSpace to see what has been updated. Advanced Queries are the only way to do that without hitting up everything, I think.

If we want to forget the Migrate API we would have to create an ArchivesSpace agent that post to the Drupal REST API on a record update.

**Comment by jasloe** [comment]
Yes, right. I'm suggesting that the agent pings the migrate job with the uri of objects that need to be updated. Could be designed to provide real-time-ish updates to D8.  

That said, I don't think we need to commit to any design decisions at the moment. Getting through authentication and some simple processing is the order of the day!

**Comment by seth-shaw-unlv** [comment]
I don't think that Migrate can take an argument like that. We shall see.

**Comment by seth-shaw-unlv** [comment]
What we could do is set the last migrate run as a Drupal State variable and then when we run the migration our source plugin can add that date (if it exists) into the search query. 

We would have to make sure that just running migrate-status doesn't reset it though, so we need to be careful about when we update it.

**Comment by seth-shaw-unlv** [comment]
FYI, I have some [initial work on importing Archival Objects](https://github.com/seth-shaw-unlv/archivesspace-drupal/tree/migrate-archival-objects).

**Comment by seth-shaw-unlv** [comment]
A major limitation of the Advanced Query method is that, at least for archival objects, it consistently reports many more archival objects than there actually are. E.g. I have a test site with 8,522 archival objects but the advanced query `{"jsonmodel_type":"advanced_query","query":{"jsonmodel_type":"boolean_query","op":"AND","subqueries":[{"field":"primary_type","value":"archival_object","comparator":"equals","jsonmodel_type":"field_query"},{"field":"user_mtime","value":"1969-12-31T18:00:00-06:00","comparator":"greater_than","jsonmodel_type":"date_field_query"}]}}` reports 12,420.

So, when I run `drush ms archival_objects` after the migration my report shows:
```
 --------------------- -------- ------- -------------- ------------- --------------------- 
  Migration ID          Status   Total   Imported       Unprocessed   Last Imported        
 --------------------- -------- ------- -------------- ------------- --------------------- 
  as_archival_objects   Idle     12420   8522 (68.6%)   3898          2019-01-16 10:01:41  
 --------------------- -------- ------- -------------- ------------- --------------------- 
```
This incorrectly leads me to believe the migration is only 68.6% complete when _it is actually done_.

So, for migrations I think we probably should just use the API for sanity's sake. We can probably find another way for triggering limited updates.
